### DISCUSSION TITLE
Option to treat all as a warning (aka eslint only-warn)

### DISCUSSION THREAD

**Discussion by fudom** [biomejs/biome#3106] [Ideas]
It would be great if we could set the linter to treat everything as warning. I can set all my rules to "warn" instead "error". But this is not possible with `recommended` rules. Otherwise I have to declare every recommended rule again, which makes no sense. 

Why only warn? 

- Don't waste time thinking or discussing about if it should be an error or a warning, focus on enabling of disabling a rule.

- Warnings look different in editors, this allows you to quickly see that some tweaking is required, but your code still runs (eslint rules generally don't block the code from executing and fatal errors are still reported as error). Often you play with code for debugging etc. Prototyping etc. Blocked by linter is annyoing. Only warn which should be fixed convention. 

Later on Git hook and/or CI pipelines we use `--error-on-warnings` (eslint: `--max-warnings=0`) because we don't want linter warnings remain. I think this is a good way how to use a linter. The developer is not blocked during coding, but forced to fix the warning before commit.

https://www.npmjs.com/package/eslint-plugin-only-warn

Implementation: 

**biome.json**

```json
{
 "linter": {
    "onlyWarn": true
  }
}
```

**Comment by Conaclos** [comment]
We could address this when we will revamp our configuration #689. Notably if we associate the severity level to the group, a user could set the severity level of every group to `warn`.

**Reply by fudom** [reply]
Ok, but this needs to cover "recommended" (external) rules. Not just your own. I'm not sure what the revamp will look like or what the plan is. But if you still set rules like "someRule": "error" etc, it could be confusing. It needs to be clear that you can set an option "onlyWarn" (or whatever it's called) that the maximum severity level will be "warn" (override). Levels below that are not affected (e.g. info).

Coming from ESLint, I was looking for a better, modern alternative as the new flat config (breaking change) was a terrible challenge. I think it will be less confusing for ESLint users if Biome uses this (warn-only) like in ESLint.

For inspiration see the tiny code: 
https://github.com/bfanger/eslint-plugin-only-warn/blob/main/src/only-warn.js

I haven't switched to Biome yet due to conflicts and config issues. Because it's a larger real project. But it looks very interessting. I will try to use it for our projects. Btw. does it have unit testing or any plant for it. Otherwise I will use Vitest. Another new insteressing thing. Currently I use Jest. But anyway... off topic. ^^

**Reply by ematipico** [reply]
There's also to note that you're referring to an eslint plugin, maintained by people that might have different opinions from eslint core.

I would like to see other tools that do that, like clippy or Ruff

**Reply by AVGVSTVS96** [reply]
> There's also to note that you're referring to an eslint plugin, maintained by people that might have different opinions from eslint core.
> 
> I would like to see other tools that do that, like clippy or Ruff

This is also available as a setting in the ESLint VSCode extension. 

```json
"eslint.rules.customizations": [{ "rule": "*", "severity": "warn" }],
```

**Reply by ematipico** [reply]
That's maintained by another party (Microsoft) that might not be related with the eslint core project.

**Comment by colelawrence** [comment]
I also want this feature. I find "error"s in VS Code highly distracting when "errors" in the IDE usually mean "this will not compile" or "this has a bug and will not work".

I would feel excellent about this being a fix for the VS Code extension side, though, since it's entirely about the editing experience for me and my team.

**Reply by AVGVSTVS96** [reply]
Agreed, I know a lot of people are fine with it as is, but I think this is an essential feature for a linter, and should be prioritized as such. 

In my opinion, this should be considered a best/standard practice to visually separate linting and LSP errors. I'm not sure why this isn't the case.

**Reply by Conaclos** [reply]
> I also want this feature. I find "error"s in VS Code highly distracting when "errors" in the IDE usually mean "this will not compile" or "this has a bug and will not work".

Currently, recommended rules emit errors and other rules emit warnings. I personally want to propose the following change for Biome 2.0: rules of the correctness group emit errors and other rules emit warnings. This should satisfy most of the requests and avoid the need for an option to turn errors into warnings. Rules from the correctness group report only invalid code.

**Reply by AVGVSTVS96** [reply]
> > I also want this feature. I find "error"s in VS Code highly distracting when "errors" in the IDE usually mean "this will not compile" or "this has a bug and will not work".
> 
> Currently, recommended rules emit errors and other rules emit warnings. I personally want to propose the following change for Biome 2.0: rules of the correctness group emit errors and other rules emit warnings. This should satisfy most of the requests and avoid the need for an option to turn errors into warnings. Rules from the correctness group report only invalid code.

That would be better, as long as we can set the Biome extension to display all lint rules as errors. 

At the same time, at this point why not make it all configurable by the user? This is pretty subjective so it would be best if it were fully configurable, maybe by rule group.

The ESLint VSC extension has this option:
```json
"eslint.rules.customizations": [{ "rule": "*", "severity": "warn" }],
```

This would be helpful, so at a glance I can tell if something is a lint rule or a real error that would stop the code from working. But it would also be nice to customize how different groups show up when running biome in the cli.

**Reply by ematipico** [reply]
I still need help understanding the use case. Why would you want to lower the severity of a rule **only in the editor**?

**Reply by jakst** [reply]
I have the same wish. The biggest reason for wanting the linter to only display warnings for me is to not confuse it with typescript errors. When I get a red squiggle my mind instantly starts racing to find out where I got the types wrong, and then I get annoyed when I see that it was the linter complaining.

While writing code I don't want to be bothered by the linter. I'll take care of that in a second pass. But seeing red squiggles makes me think my code will crash when I run it (this is not always true for Typescript ofc, but it mostly is).

**Reply by jakst** [reply]
Basically this: Let me know there are lint errors, but I wanna continue coding for a while before I take care of it, so don't scream at me.

**Reply by ematipico** [reply]
@jakst  That doesn't answer my question. The setting that people are using as a reference **is for editors only**, which means that the CLI (eslint) will result in a different behaviour than the editor. I want to understand why people want to lower the severity of lint rules **_only in the editors_**, despite having the CLI behaving differently.

**Reply by jakst** [reply]
The reason you'd want it to be controllable in the editor separately is because that's where it really gets in your face. If I run a lint command from the CLI I don't mind as much that they're listed as errors, because I'm in a mode of looking for lint errors. But in the editor I have no choice but to look at them, even if that's not my focus atm.

For myself I'd still prefer to be able to set all rules to "warn" by default in the biome config, but you might not always be in control of that in a larger org. In those cases it's very nice to be able to turn down the severity in your editor for the reasons listed above.

**Reply by colelawrence** [reply]
@ematipico the reason it's bad in the editor for me is that there's no way to differentiate between linting errors and type errors. Type errors actually mean "this code probably has a bug so you can't run it" whereas linting errors mean "here's something to fix before committing".

If I could style and separate the linter's errors from my type checker's errors (visually and in the "Problems" pane), then I would be super happy, but that isn't possible.

**Reply by net** [reply]
I don't think most lint failures should be displayed as errors in the CLI either. You can and should set up your CI so it fails on lint warnings (`--error-on-warnings`).

If there's a rule whose failure you want to allow, it should be either selectively disabled with comments, or its level should be `info` (not `warning`).

This doesn't mean there isn't a place in a linter for actual errors though; there are rare cases where a linter can catch runtime errors that the compiler isn't catching (like `noConstAssign` and the other rules in the [correctness](https://next.biomejs.dev/linter/javascript/rules/#correctness) group), and those should show up as actual errors in _both_ the editor and CLI.

**Comment by pedramardakani** [comment]
OK, I know this is definitely nothing more than an ugly temporary solution, but some might find it helpful:

## TL;DR

Copy this to reduce severity of most of the rules as `warn`

```json
{
  "$schema": "https://biomejs.dev/schemas/1.8.3/schema.json",
  "formatter": { "enabled": false },
  "organizeImports": { "enabled": true },
  "linter": {
    "enabled": true,
    "rules": {
      "complexity": {
        "noUselessSwitchCase": "warn",
        "noUselessFragments": "warn",
        "noThisInStatic": "warn",
        "useSimpleNumberKeys": "warn",
        "noUselessCatch": "warn",
        "noEmptyTypeParameters": "warn",
        "noUselessThisAlias": "warn",
        "noUselessConstructor": "warn",
        "noUselessLoneBlockStatements": "warn",
        "useFlatMap": "warn",
        "useRegexLiterals": "warn",
        "noStaticOnlyClass": "warn",
        "useOptionalChain": "warn",
        "noUselessEmptyExport": "warn",
        "noExtraBooleanCast": "warn",
        "noUselessTypeConstraint": "warn",
        "noForEach": "warn",
        "noBannedTypes": "warn",
        "noMultipleSpacesInRegularExpressionLiterals": "warn",
        "noWith": "warn",
        "useLiteralKeys": "warn",
        "noUselessTernary": "warn",
        "noUselessRename": "warn",
        "useArrowFunction": "warn",
        "noExcessiveNestedTestSuites": "warn",
        "noUselessLabel": "warn"
      },
      "correctness": {
        "noUnreachable": "warn",
        "noSetterReturn": "warn",
        "useYield": "warn",
        "noSwitchDeclarations": "warn",
        "noUnusedLabels": "warn",
        "noUnsafeFinally": "warn",
        "noChildrenProp": "warn",
        "useIsNan": "warn",
        "noEmptyPattern": "warn",
        "noInvalidConstructorSuper": "warn",
        "noRenderReturnValue": "warn",
        "noSelfAssign": "warn",
        "noConstAssign": "warn",
        "noFlatMapIdentity": "warn",
        "noStringCaseMismatch": "warn",
        "noPrecisionLoss": "warn",
        "noGlobalObjectCalls": "warn",
        "noUnreachableSuper": "warn",
        "noConstructorReturn": "warn",
        "noNonoctalDecimalEscape": "warn",
        "useJsxKeyInIterable": "warn",
        "noEmptyCharacterClassInRegex": "warn",
        "noInnerDeclarations": "warn",
        "noConstantCondition": "warn",
        "noVoidElementsWithChildren": "warn",
        "noUnsafeOptionalChaining": "warn",
        "useExhaustiveDependencies": "warn",
        "useValidForDirection": "warn",
        "noInvalidUseBeforeDeclaration": "warn",
        "noUnnecessaryContinue": "warn",
        "noVoidTypeReturn": "warn"
      },
      "nursery": {},
      "performance": { "noDelete": "warn", "noAccumulatingSpread": "warn" },
      "security": {
        "noDangerouslySetInnerHtmlWithChildren": "warn",
        "noGlobalEval": "warn",
        "noDangerouslySetInnerHtml": "warn"
      },
      "style": {
        "useSelfClosingElements": "warn",
        "noCommaOperator": "warn",
        "noInferrableTypes": "warn",
        "useConst": "warn",
        "useShorthandFunctionType": "warn",
        "noArguments": "warn",
        "useDefaultParameterLast": "warn",
        "noUnusedTemplateLiteral": "warn",
        "useExponentiationOperator": "warn",
        "useEnumInitializers": "warn",
        "useAsConstAssertion": "warn",
        "useNodejsImportProtocol": "warn",
        "useNumberNamespace": "warn",
        "noNonNullAssertion": "warn",
        "useWhile": "warn",
        "useLiteralEnumMembers": "warn",
        "noUselessElse": "warn",
        "useSingleVarDeclarator": "warn",
        "noParameterAssign": "warn",
        "useNumericLiterals": "warn",
        "useExportType": "warn",
        "useImportType": "warn",
        "noVar": "warn",
        "useTemplate": "warn"
      },
      "suspicious": {
        "noAsyncPromiseExecutor": "warn",
        "noLabelVar": "warn",
        "noDuplicateObjectKeys": "warn",
        "noUnsafeDeclarationMerging": "warn",
        "noDoubleEquals": "warn",
        "noDuplicateParameters": "warn",
        "noFocusedTests": "warn",
        "noFunctionAssign": "warn",
        "noGlobalIsNan": "warn",
        "noConfusingVoidType": "warn",
        "useDefaultSwitchClauseLast": "warn",
        "useIsArray": "warn",
        "noImplicitAnyLet": "warn",
        "noEmptyInterface": "warn",
        "noExplicitAny": "warn",
        "noAssignInExpressions": "warn",
        "useGetterReturn": "warn",
        "useValidTypeof": "warn",
        "useNamespaceKeyword": "warn",
        "noControlCharactersInRegex": "warn",
        "noMisleadingCharacterClass": "warn",
        "noImportAssign": "warn",
        "noExtraNonNullAssertion": "warn",
        "noCommentText": "warn",
        "noFallthroughSwitchClause": "warn",
        "noShadowRestrictedNames": "warn",
        "noArrayIndexKey": "warn",
        "noRedeclare": "warn",
        "noConfusingLabels": "warn",
        "noCompareNegZero": "warn",
        "noConstEnum": "warn",
        "noMisrefactoredShorthandAssign": "warn",
        "noCatchAssign": "warn",
        "noSelfCompare": "warn",
        "noGlobalAssign": "warn",
        "noPrototypeBuiltins": "warn",
        "noClassAssign": "warn",
        "noDuplicateCase": "warn",
        "noApproximativeNumericConstant": "warn",
        "noDebugger": "warn",
        "noGlobalIsFinite": "warn",
        "noDuplicateTestHooks": "warn",
        "noMisleadingInstantiator": "warn",
        "noRedundantUseStrict": "warn",
        "noThenProperty": "warn",
        "noDuplicateClassMembers": "warn",
        "noSuspiciousSemicolonInJsx": "warn",
        "noUnsafeNegation": "warn",
        "noSparseArray": "warn",
        "noDuplicateJsxProps": "warn",
        "noExportsInTest": "warn"
      }
    }
  }
}
```

## Complete but broken version

This one is generated based on the ones already there on biomejs website. However, the server would complain that the json is misconfigured - some unknown keywords 🤷‍♀️

```json
{
  "$schema": "https://biomejs.dev/schemas/1.8.3/schema.json",
  "formatter": { "enabled": false },
  "organizeImports": { "enabled": true },
  "linter": {
    "enabled": true,
    "rules": {
      "accessibility": {
        "useMediaCaption": "warn",
        "noAutofocus": "warn",
        "noRedundantAlt": "warn",
        "noLabelWithoutControl": "warn",
        "useHtmlLang": "warn",
        "useAriaActivedescendantWithTabindex": "warn",
        "noAriaHiddenOnFocusable": "warn",
        "useIframeTitle": "warn",
        "noAccessKey": "warn",
        "noDistractingElements": "warn",
        "noRedundantRoles": "warn",
        "noNoninteractiveElementToInteractiveRole": "warn",
        "useAriaPropsForRole": "warn",
        "useGenericFontNames": "warn",
        "useFocusableInteractive": "warn",
        "noBlankTarget": "warn",
        "useValidAnchor": "warn",
        "noHeaderScope": "warn",
        "useValidLang": "warn",
        "useKeyWithMouseEvents": "warn",
        "noNoninteractiveTabindex": "warn",
        "useAltText": "warn",
        "noSvgWithoutTitle": "warn",
        "useHeadingContent": "warn",
        "noPositiveTabindex": "warn",
        "useAnchorContent": "warn",
        "useValidAriaRole": "warn",
        "useValidAriaValues": "warn",
        "useSemanticElements": "warn",
        "useButtonType": "warn",
        "useValidAriaProps": "warn",
        "useKeyWithClickEvents": "warn",
        "noAriaUnsupportedElements": "warn",
        "noInteractiveElementToNoninteractiveRole": "warn"
      },
      "complexity": {
        "noUselessSwitchCase": "warn",
        "noUselessFragments": "warn",
        "noThisInStatic": "warn",
        "noUselessUndefinedInitialization": "warn",
        "useSimpleNumberKeys": "warn",
        "noUselessCatch": "warn",
        "noEmptyTypeParameters": "warn",
        "noUselessThisAlias": "warn",
        "noVoid": "warn",
        "noUselessConstructor": "warn",
        "noUselessLoneBlockStatements": "warn",
        "useFlatMap": "warn",
        "noExcessiveCognitiveComplexity": "warn",
        "useRegexLiterals": "warn",
        "useDateNow": "warn",
        "noStaticOnlyClass": "warn",
        "useOptionalChain": "warn",
        "noUselessEmptyExport": "warn",
        "noExtraBooleanCast": "warn",
        "noUselessTypeConstraint": "warn",
        "noForEach": "warn",
        "noBannedTypes": "warn",
        "noMultipleSpacesInRegularExpressionLiterals": "warn",
        "noWith": "warn",
        "useLiteralKeys": "warn",
        "noUselessTernary": "warn",
        "noUselessRename": "warn",
        "noUselessStringConcat": "warn",
        "useSimplifiedLogicExpression": "warn",
        "useArrowFunction": "warn",
        "noExcessiveNestedTestSuites": "warn",
        "noUselessLabel": "warn"
      },
      "correctness": {
        "noUnreachable": "warn",
        "noInvalidBuiltinInstantiation": "warn",
        "noUnknownMediaFeatureName": "warn",
        "noNodejsModules": "warn",
        "noUnknownUnit": "warn",
        "noSetterReturn": "warn",
        "useImportExtensions": "warn",
        "noInvalidPositionAtImportRule": "warn",
        "useYield": "warn",
        "noSwitchDeclarations": "warn",
        "noUnknownFunction": "warn",
        "noUnusedLabels": "warn",
        "noUnusedPrivateClassMembers": "warn",
        "noUnsafeFinally": "warn",
        "noChildrenProp": "warn",
        "noInvalidNewBuiltin": "warn",
        "useIsNan": "warn",
        "noEmptyPattern": "warn",
        "noInvalidConstructorSuper": "warn",
        "noInvalidDirectionInLinearGradient": "warn",
        "noRenderReturnValue": "warn",
        "noUnknownProperty": "warn",
        "noUnusedFunctionParameters": "warn",
        "noInvalidGridAreas": "warn",
        "noSelfAssign": "warn",
        "useArrayLiterals": "warn",
        "noConstAssign": "warn",
        "noFlatMapIdentity": "warn",
        "noStringCaseMismatch": "warn",
        "noUnusedImports": "warn",
        "noUndeclaredVariables": "warn",
        "noPrecisionLoss": "warn",
        "noGlobalObjectCalls": "warn",
        "noUnreachableSuper": "warn",
        "noConstructorReturn": "warn",
        "noNonoctalDecimalEscape": "warn",
        "useJsxKeyInIterable": "warn",
        "noEmptyCharacterClassInRegex": "warn",
        "noInnerDeclarations": "warn",
        "noConstantCondition": "warn",
        "noVoidElementsWithChildren": "warn",
        "noUnsafeOptionalChaining": "warn",
        "useExhaustiveDependencies": "warn",
        "useValidForDirection": "warn",
        "noUndeclaredDependencies": "warn",
        "noInvalidUseBeforeDeclaration": "warn",
        "noUnnecessaryContinue": "warn",
        "useHookAtTopLevel": "warn",
        "noVoidTypeReturn": "warn",
        "noUnmatchableAnbSelector": "warn",
        "noUnusedVariables": "warn",
        "noNewSymbol": "warn",
        "noConstantMathMinMaxClamp": "warn"
      },
      "nursery": {
        "useValidAutocomplete": "warn",
        "useAdjacentOverloadSignatures": "warn",
        "useDeprecatedReason": "warn",
        "noUnknownPseudoClass": "warn",
        "useStrictMode": "warn",
        "noUselessEscapeInRegex": "warn",
        "useConsistentCurlyBraces": "warn",
        "noDuplicatedFields": "warn",
        "noUnknownPseudoElement": "warn",
        "noExportedImports": "warn",
        "useSortedClasses": "warn",
        "noIrregularWhitespace": "warn",
        "noValueAtRule": "warn",
        "noRestrictedImports": "warn",
        "noProcessEnv": "warn",
        "noSecrets": "warn",
        "useConsistentMemberAccessibility": "warn",
        "useTrimStartEnd": "warn",
        "noCommonJs": "warn",
        "useAriaPropsSupportedByRole": "warn",
        "noDynamicNamespaceImportAccess": "warn",
        "noDuplicateCustomProperties": "warn",
        "noStaticElementInteractions": "warn",
        "noRestrictedTypes": "warn",
        "noDuplicateElseIf": "warn",
        "noEnum": "warn",
        "useImportRestrictions": "warn",
        "noSubstr": "warn"
      },
      "performance": {
        "noBarrelFile": "warn",
        "useTopLevelRegex": "warn",
        "noDelete": "warn",
        "noAccumulatingSpread": "warn",
        "noReExportAll": "warn"
      },
      "security": {
        "noDangerouslySetInnerHtmlWithChildren": "warn",
        "noGlobalEval": "warn",
        "noDangerouslySetInnerHtml": "warn"
      },
      "style": {
        "useSelfClosingElements": "warn",
        "noRestrictedGlobals": "warn",
        "useConsistentBuiltinInstantiation": "warn",
        "noCommaOperator": "warn",
        "noInferrableTypes": "warn",
        "noNamespaceImport": "warn",
        "useExplicitLengthCheck": "warn",
        "useFilenamingConvention": "warn",
        "useConst": "warn",
        "useShorthandFunctionType": "warn",
        "noArguments": "warn",
        "useDefaultParameterLast": "warn",
        "useFragmentSyntax": "warn",
        "noUnusedTemplateLiteral": "warn",
        "useExponentiationOperator": "warn",
        "useEnumInitializers": "warn",
        "useAsConstAssertion": "warn",
        "useNamingConvention": "warn",
        "noDoneCallback": "warn",
        "useNodejsImportProtocol": "warn",
        "useThrowNewError": "warn",
        "noNamespace": "warn",
        "noShoutyConstants": "warn",
        "useNumberNamespace": "warn",
        "noNonNullAssertion": "warn",
        "useForOf": "warn",
        "useBlockStatements": "warn",
        "noYodaExpression": "warn",
        "useCollapsedElseIf": "warn",
        "useDefaultSwitchClause": "warn",
        "useShorthandAssign": "warn",
        "useWhile": "warn",
        "useLiteralEnumMembers": "warn",
        "useShorthandArrayType": "warn",
        "noUselessElse": "warn",
        "noImplicitBoolean": "warn",
        "useNodeAssertStrict": "warn",
        "useThrowOnlyError": "warn",
        "useSingleVarDeclarator": "warn",
        "useSingleCaseStatement": "warn",
        "noParameterAssign": "warn",
        "noParameterProperties": "warn",
        "useNumericLiterals": "warn",
        "useExportType": "warn",
        "useImportType": "warn",
        "noVar": "warn",
        "useConsistentArrayType": "warn",
        "useTemplate": "warn",
        "noNegationElse": "warn",
        "noDefaultExport": "warn"
      },
      "suspicious": {
        "noAsyncPromiseExecutor": "warn",
        "noLabelVar": "warn",
        "noDuplicateAtImportRules": "warn",
        "noDuplicateObjectKeys": "warn",
        "noUnsafeDeclarationMerging": "warn",
        "noConsole": "warn",
        "noDoubleEquals": "warn",
        "noDuplicateParameters": "warn",
        "noFocusedTests": "warn",
        "noFunctionAssign": "warn",
        "noGlobalIsNan": "warn",
        "noSkippedTests": "warn",
        "noConfusingVoidType": "warn",
        "useAwait": "warn",
        "useDefaultSwitchClauseLast": "warn",
        "useIsArray": "warn",
        "noDuplicateSelectorsKeyframeBlock": "warn",
        "noConsoleLog": "warn",
        "noImplicitAnyLet": "warn",
        "noEmptyInterface": "warn",
        "noExplicitAny": "warn",
        "noAssignInExpressions": "warn",
        "noMisplacedAssertion": "warn",
        "useGetterReturn": "warn",
        "useValidTypeof": "warn",
        "useNamespaceKeyword": "warn",
        "useErrorMessage": "warn",
        "noControlCharactersInRegex": "warn",
        "noMisleadingCharacterClass": "warn",
        "useNumberToFixedDigitsArgument": "warn",
        "noImportAssign": "warn",
        "noExtraNonNullAssertion": "warn",
        "noImportantInKeyframe": "warn",
        "noCommentText": "warn",
        "noFallthroughSwitchClause": "warn",
        "noShadowRestrictedNames": "warn",
        "noArrayIndexKey": "warn",
        "noRedeclare": "warn",
        "noConfusingLabels": "warn",
        "noCompareNegZero": "warn",
        "noConstEnum": "warn",
        "noMisrefactoredShorthandAssign": "warn",
        "noCatchAssign": "warn",
        "noSelfCompare": "warn",
        "noEmptyBlockStatements": "warn",
        "noGlobalAssign": "warn",
        "noPrototypeBuiltins": "warn",
        "noClassAssign": "warn",
        "noDuplicateCase": "warn",
        "noApproximativeNumericConstant": "warn",
        "noDebugger": "warn",
        "noGlobalIsFinite": "warn",
        "noDuplicateTestHooks": "warn",
        "noMisleadingInstantiator": "warn",
        "noEvolvingTypes": "warn",
        "noRedundantUseStrict": "warn",
        "noThenProperty": "warn",
        "noDuplicateClassMembers": "warn",
        "noEmptyBlock": "warn",
        "noReactSpecificProps": "warn",
        "noSuspiciousSemicolonInJsx": "warn",
        "noUnsafeNegation": "warn",
        "noDuplicateFontNames": "warn",
        "noSparseArray": "warn",
        "noDuplicateJsxProps": "warn",
        "noExportsInTest": "warn",
        "noShorthandPropertyOverrides": "warn"
      }
    }
  }
}

```

## How?

Use the following python code to generate the list yourselves - it needs beautiful soup 4 as a dependency:

```python
from bs4 import BeautifulSoup
import requests as rq
request = rq.get("https://biomejs.dev/linter/rules/")
page = BeautifulSoup(request.content)
def rules_set(name: str):
    rules = page.find(id=name).parent.find_next_sibling("table")
    return set([item.text for item in rules.find_all("td")][::3])

rule_groups = ["accessibility", "complexity", "correctness", "nursery", "performance", "security", "style", "suspicious"]
recommended_rules = page.find(id="recommended-rules").parent.find_next_sibling("ul")
recommended_rules_set = set(item.text for item in recommended_rules)
recommended_rules_dict = {r: {s: "warn" for s in rules_set(r) if s in recommended_rules_set} for r in rule_groups}

# Only recommended rules
import json
print(json.dumps(recommended_rules_dict, indent=2))

## If you want to see all the rules
everything = {r: {s: "warn" for s in rules_set(r)} for r in rule_groups}
print(json.dumps(everything, indent=2))
```

**Comment by niedzielski** [comment]
big +1! I'd like to configure these stylistic rules as warnings so that they appear as friendly yellow squigglies I know I can ignore during development and can defer fixing until commit time. I end up wasting time fixing stylistic issues during development instead of building something functional because my legitimate program-breaking bugs are mixed up with red stylistic errors. I forbid warnings in my codebase at commit time and that's the only time I care about style.

**Comment by net** [comment]
I strongly urge that this behavior actually be the default out of the box behavior. Errors should **only be for actual errors** that will prevent compilation or will provably error during runtime; **stylistic lints should always show up as warnings**.

ESLint got this wrong, Biome has a chance to get this right out of the box. ❤️

**Reply by ematipico** [reply]
style rules are now set to warnings by default, unless configured otherwise. This was changed in v2

**Reply by net** [reply]
Unless I'm missing something, I still see a lot of rules that aren't compilation or runtime errors whose default severity is error. Example:

https://next.biomejs.dev/linter/rules/use-flat-map/

https://next.biomejs.dev/linter/rules/no-extra-boolean-cast/

The reason **only** compilation or runtime errors should be displayed as errors is to allow the developer to easily iterate on work-in-progress code without having to fix stylistic lints like `use-flat-map` and `no-extra-boolean-cast`, then fix those lints at the end before opening a PR or merging their changes.

Errors should tell them what they need to fix before their code will run; warnings tell them what they need to fix before their code is mergable.

This also avoids bikeshedding over which lints should be errors and which should be warnings. 

This is why the popular `eslint-plugin-only-warn` package sets _all_ lints to warnings, instead of just some.

**Reply by ematipico** [reply]
> Unless I'm missing something

Yes, in my previous comment I said **style** rules. It means rules that belong to the **style** group. The two rules that you shared belong to the complexity group.

**Reply by net** [reply]
@ematipico I see.

I'm proposing that nearly all rules default to warning instead of error, _including_ complexity rules like `useFlatMap`.

There are rules that should still be errors by default, like `noConstAssign`, because those rules detect actual runtime errors (and so have to be resolved before code can be executed). I believe every rule in the correctness group is a true error. A rule like `useFlatMap` doesn't detect a runtime error, and so does not need to be fixed before executing code during development (but should be fixed before merging your code, if that is your team's preference).

To recap, ESLint encouraged the convention:

- **error**: must be fixed before merging (this causes the issues that this discussion is about in that non-error rules like show up as errors in editors)
- **warning**: shown but allowed to be merged
- **off**: disabled

Biome has a chance to default to something much better:

- **error**: must be fixed before running
- **warning**: must be fixed before merging
- **info**: shown but allowed to be merged
- **off**: disabled

(and then `--error-on-warnings` should also be the default)

**Comment by ematipico** [comment]
This is now possible in v2

**Reply by jakst** [reply]
The v2 beta announcement doesn't mention anything about this. Just that style rules aren't recommended anymore. How does one switch biome to treat all rules as warnings in v2?

**Reply by ematipico** [reply]
The announcement only mentions a glimpse of all the features we have.

I suggest you browse the new docs or check the changelog.

https://next.biomejs.dev/linter/#group-severity

**Reply by jakst** [reply]
Very nice! 👏

**Reply by net** [reply]
@ematipico it's not clear to me how that can be used without enabling rules that aren't recommended. I tried

```json
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
      "complexity":"warn",
      "performance": "warn",
      "security": "warn",
      "style": "warn",
      "suspicious": "warn"
    }
  },
```

but this seems to enable _all_ rules.

**Reply by ematipico** [reply]
I suppose you are right. It isn't the same thing.

Maybe having a `linter.severity` option could do the trick

**Reply by boehs** [reply]
> I suppose you are right. It isn't the same thing.
> 
> Maybe having a `linter.severity` option could do the trick

please! that would be huge
